c/make-derivation.nix

```diff
@@ -0,0 +1,191 @@
+
+        lintPhase = attrs.lintPhase or ''
+          runHook preLint
+          if [ -z "''${dontCheckClangFormat:-}" ]; then
+            echo "🏟 Checking format in C/C++ files..."
+            ${buildPackages.fd}/bin/fd --ignore-file=.gitignore --glob '*.[h,hpp,hh,cpp,cxx,cc,c]' --exec-batch clang-format -Werror -n --style=file
+            rc=$?
+
+            if [ $rc -eq 0 ]; then
+              echo "✅ perfectly formatted!"
+            fi
+          fi
+          runHook postLint
+        '';
+
+        shellCommands = {
+          build = {
+            # TODO: fix undef of buildPhase when we stop using mkShell
+            script = ''
+              : ''${NIX_LOG_FD:=} ''${buildFlags:=}
+              ${if attrs ? buildPhase then "" else "unset buildPhase"}
+              phases="''${preConfigurePhases:-} configurePhase ''${preBuildPhases:-} buildPhase"
+              genericBuild
+            '';
+            description = "Build the component.";
+          };
+          format = {
+            script = ''
+              runHook preFormat
+              echo "🏟️ Formatting C++ files..."
+              ${buildPackages.fd}/bin/fd --glob '*.[h,hpp,hh,cpp,cxx,cc,c]' --exec-batch clang-format --style=file -i "$@"
+              runHook postFormat
+            '';
+            description = "Run clang-format in the current component.";
+          };
+          check = {
+            script = ''
+              : ''${NIX_LOG_FD:=} ''${buildFlags:=} ''${lintPhase:=echo 'no lintPhase'}
+              ${if attrs ? buildPhase then "" else "unset buildPhase"}
+              phases="''${preConfigurePhases:-} configurePhase ''${preBuildPhases:-} buildPhase checkPhase lintPhase"
+              genericBuild
+            '';
+            description = "Run the checks and lints for the component.";
+          };
+          debug = {
+            script = ''${attrs.debugger or "gdb"} ${if attrs ? executable then "${attrs.executable} \"$@\"" else "$@"}'';
+            description = "Run ${attrs.debugger or "gdb"}${if attrs ? executable then " on ${attrs.executable}" else ""}.";
+          };
+        }
+        // lib.optionalAttrs (attrs ? executable) {
+          run = {
+            script = ''./${attrs.executable} "$@"'';
+            description = "Run ${attrs.executable}";
+          };
+        } // lib.optionalAttrs enableDoxygen {
+          show-doxyfile = {
+            script = "doxygen --print-generated-config";
+            description = "Show the combination of Doxygen files that will be used by Doxygen.";
+          };
+          docs = {
+            script = ''
+              doxygen && xdg-open ${attrs.doxygenOutputDir or doxygenOutputDir}/html/index.html
+            '';
+            description = "Show the documentation.";
+          };
+        }
+        // attrs.shellCommands or { };
+      }));
+
+  targetComponents = base.mapComponentsRecursive
+    (_path: component:
+      (component."${targetName}" or component) // {
```

Comment:
Shouldn't this be `component.${targetName} or component._default`? And maybe `or component` at the end but I think I would prefer an error

Reply:
Not all things have targets when we recurse through it. For example component sets. We still need to go through them and check what's inside them.

Reply:
Think we just want the target they asked for. Example:
```
A {
  inputs = [ B ];
  targets = {
    windows = ...
    _default = ...
  };
}

B {
...
  targets = {
    _default = ...
  };
...
}
```

Component A depends on Component B. Component A can be built with windows and default.
Component B does not have a target for windows. If we want to build A for windows wouldn't it be wrong to take the default target for B?

Either way I noticed it's not working as expected. Will have another look into it.

Reply:
For rust we look for the "rust" target if we can't find the one we are, we don't have a "c" target so my thinking was that _default was the most similar thing. When accessing something compiled you'll very likely get problems, especially between nix and windows, but sometimes you might be doing things between platforms with less difference and sometimes you might want to access other files than binaries or properties on the derivation.

Reply:
`targetName` will be `_default` in the only case that makes sense anyway so no need to look in more places, will not be correct or provide anything additional

Reply:
And in the case you want something special, on another platform, you can just use the component as normal and pick out the target. Assuming people want to use the `_default` might not be correct.

Reply:
> but sometimes you might be doing things between platforms with less difference and sometimes you might want to access other files than binaries or properties on the derivation.

True dat, will see if I can come up with something for that

Reply:
> > but sometimes you might be doing things between platforms with less difference and sometimes you might want to access other files than binaries or properties on the derivation.
> 
> True dat, will see if I can come up with something for that

You can always depend on a component on the top and pick and choose whatever, instead of depending on it in the mk-function so I don't think we need any new mechanic.

Reply:
> You can always depend on a component on the top and pick and choose whatever, instead of depending on it in the mk-function so I don't think we need any new mechanic.

True, although I did come up with another thing :)

Reply:
It's a nice solution, we have as many solutions to this as we have languages now so at some point we should consolidate.